File: java/src/main/java/lab/gabon/service/AuthService.java
package lab.gabon.service;

import lab.gabon.common.AppError;
import lab.gabon.common.AppException;
import lab.gabon.model.entity.Customer;
import lab.gabon.model.response.AuthResponses.CustomerMeResponse;
import lab.gabon.model.response.AuthResponses.TokenPairResponse;
import lab.gabon.repository.CustomerRepository;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;

@Service
public class AuthService {

  private final CustomerRepository customerRepo;
  private final JwtService jwtService;
  private final TokenStore tokenStore;
  private final BCryptPasswordEncoder encoder = new BCryptPasswordEncoder(10);

  public AuthService(
      CustomerRepository customerRepo, JwtService jwtService, TokenStore tokenStore) {
    this.customerRepo = customerRepo;
    this.jwtService = jwtService;
    this.tokenStore = tokenStore;
  }

  public TokenPairResponse register(String username, String password) {
    customerRepo
        .findByUsername(username)
        .ifPresent(
            c -> {
              throw new AppException(new AppError.UsernameExists());
            });

    var hash = encoder.encode(password);
    var customer =
        new Customer(
            null, username, hash, null, null, null, null, null, false, 0, null, null, null, null,
            null);
    var saved = customerRepo.save(customer);

    var tokens = jwtService.generateCustomerTokens(saved.id(), null);
    tokenStore.storeRefreshFamily(tokens.familyId(), saved.id(), tokens.refreshJti(), false);

    return new TokenPairResponse(tokens.accessToken(), tokens.refreshToken());
  }

  public TokenPairResponse login(String username, String password) {
    var customer =
        customerRepo
            .findByUsername(username)
            .orElseThrow(() -> new AppException(new AppError.InvalidCredentials()));

    if (!encoder.matches(password, customer.passwordHash())) {
      throw new AppException(new AppError.InvalidCredentials());
    }

    customerRepo.updateLastLogin(customer.id());

    var tokens = jwtService.generateCustomerTokens(customer.id(), null);
    tokenStore.storeRefreshFamily(tokens.familyId(), customer.id(), tokens.refreshJti(), false);

    return new TokenPairResponse(tokens.accessToken(), tokens.refreshToken());
  }

  public TokenPairResponse refresh(String refreshToken) {
    var claims = jwtService.verifyCustomerRefresh(refreshToken);

    var newTokens = jwtService.generateCustomerTokens(claims.userId(), claims.familyId());

    var casResult =
        tokenStore.rotateRefreshToken(claims.familyId(), claims.jti(), newTokens.refreshJti());

    return switch (casResult) {
      case TokenStore.CasResult.Success s ->
          new TokenPairResponse(newTokens.accessToken(), newTokens.refreshToken());
      case TokenStore.CasResult.Missing m -> throw new AppException(new AppError.TokenInvalid());
      case TokenStore.CasResult.Conflict c -> throw new AppException(new AppError.TokenInvalid());
    };
  }

  public void logout(String accessToken, String refreshToken) {
    var accessClaims = jwtService.verifyCustomerAccess(accessToken);
    tokenStore.blacklistAccessToken(accessClaims.jti(), false);

    try {
      var refreshClaims = jwtService.verifyCustomerRefresh(refreshToken);
      tokenStore.revokeFamily(refreshClaims.familyId());
    } catch (AppException e) {
      // Ignore errors if refresh token is expired/invalid
    }
  }

  public CustomerMeResponse getMe(long customerId) {
    var customer =
        customerRepo
            .findActiveById(customerId)
            .orElseThrow(() -> new AppException(new AppError.NotFound("customer not found")));

    return toMeResponse(customer);
  }

  public void updatePassword(long customerId, String currentPassword, String newPassword) {
    var customer =
        customerRepo
            .findActiveById(customerId)
            .orElseThrow(() -> new AppException(new AppError.NotFound("customer not found")));

    if (!encoder.matches(currentPassword, customer.passwordHash())) {
      throw new AppException(new AppError.PasswordMismatch());
    }

    customerRepo.updatePassword(customerId, encoder.encode(newPassword));
  }

  private static CustomerMeResponse toMeResponse(Customer c) {
    return new CustomerMeResponse(
        c.id(),
        c.username(),
        c.name(),
        c.phone(),
        c.email(),
        c.avatarUrl(),
        c.signature(),
        c.isVip(),
        c.diamondBalance());
  }
}

File: kotlin/src/main/kotlin/lab/gabon/service/AuthService.kt
package lab.gabon.service

import at.favre.lib.crypto.bcrypt.BCrypt
import lab.gabon.model.AppError
import lab.gabon.model.AppException
import lab.gabon.repository.CustomerRepo
import lab.gabon.repository.CustomerRow

data class TokenResponse(
    val accessToken: String,
    val refreshToken: String,
)

data class CustomerProfile(
    val id: Long,
    val username: String,
    val name: String?,
    val phone: String?,
    val email: String?,
    val avatarUrl: String?,
    val signature: String?,
    val isVip: Boolean,
    val diamondBalance: Long,
)

class AuthService(
    private val customerRepo: CustomerRepo,
    private val jwtService: JwtService,
    private val tokenStore: RedisTokenStore,
) {
    suspend fun register(
        username: String,
        password: String,
    ): TokenResponse {
        validateInput(username, password)

        val existing = customerRepo.findByUsername(username)
        if (existing != null) {
            throw AppException(AppError.UsernameExists())
        }

        val hash = hashPassword(password)
        val customerId = customerRepo.create(username, hash)

        val tokenPair = jwtService.generateCustomerTokens(customerId)
        tokenStore.setFamily(
            familyId = tokenPair.familyId,
            userId = customerId,
            currentJti = tokenPair.refreshJti,
            ttlSeconds = jwtService.config.customerRefreshTtl.inWholeSeconds,
        )

        return TokenResponse(tokenPair.accessToken, tokenPair.refreshToken)
    }

    suspend fun login(
        username: String,
        password: String,
    ): TokenResponse {
        val customer =
            customerRepo.findByUsername(username)
                ?: throw AppException(AppError.InvalidCredentials())

        if (!verifyPassword(password, customer.passwordHash)) {
            throw AppException(AppError.InvalidCredentials())
        }

        customerRepo.updateLastLogin(customer.id)

        val tokenPair = jwtService.generateCustomerTokens(customer.id)
        tokenStore.setFamily(
            familyId = tokenPair.familyId,
            userId = customer.id,
            currentJti = tokenPair.refreshJti,
            ttlSeconds = jwtService.config.customerRefreshTtl.inWholeSeconds,
        )

        return TokenResponse(tokenPair.accessToken, tokenPair.refreshToken)
    }

    suspend fun refresh(refreshTokenStr: String): TokenResponse {
        val claims = jwtService.parseCustomerToken(refreshTokenStr)

        if (claims.tokenType != "refresh") {
            throw AppException(AppError.TokenInvalid("not a refresh token"))
        }

        // Generate new token pair first, then atomically CAS the family
        val newTokenPair = jwtService.generateCustomerTokens(claims.userId)

        val casResult =
            tokenStore.casFamily(
                familyId = claims.familyId,
                expectedJti = claims.jti,
                newJti = newTokenPair.refreshJti,
            )

        return when (casResult) {
            is CasResult.Success ->
                TokenResponse(newTokenPair.accessToken, newTokenPair.refreshToken)
            is CasResult.Missing ->
                throw AppException(AppError.TokenInvalid("token family expired or revoked"))
            is CasResult.Conflict ->
                throw AppException(AppError.TokenInvalid("token reuse detected, family revoked"))
        }
    }

    @Suppress("UnusedParameter")
    suspend fun logout(
        customerId: Long,
        jti: String,
        familyId: String,
    ) {
        val remainingSeconds = jwtService.config.customerAccessTtl.inWholeSeconds
        tokenStore.setBlacklist(jti, remainingSeconds)
        tokenStore.deleteFamily(familyId)
    }

    suspend fun changePassword(
        customerId: Long,
        oldPassword: String,
        newPassword: String,
    ) {
        val customer =
            customerRepo.findById(customerId)
                ?: throw AppException(AppError.NotFound("customer not found"))

        if (!verifyPassword(oldPassword, customer.passwordHash)) {
            throw AppException(AppError.PasswordMismatch())
        }

        val newHash = hashPassword(newPassword)
        customerRepo.updatePassword(customerId, newHash)
    }

    suspend fun getMe(customerId: Long): CustomerProfile {
        val customer =
            customerRepo.findById(customerId)
                ?: throw AppException(AppError.NotFound("customer not found"))

        return customer.toProfile()
    }

    private fun validateInput(
        username: String,
        password: String,
    ) {
        if (username.length < MIN_USERNAME_LENGTH) {
            throw AppException(AppError.BadRequest("username must be at least $MIN_USERNAME_LENGTH characters"))
        }
        if (password.length < MIN_PASSWORD_LENGTH) {
            throw AppException(AppError.BadRequest("password must be at least $MIN_PASSWORD_LENGTH characters"))
        }
    }

    private fun hashPassword(password: String): String =
        BCrypt
            .withDefaults()
            .hashToString(BCRYPT_COST, password.toCharArray())

    private fun verifyPassword(
        password: String,
        hash: String,
    ): Boolean = BCrypt.verifyer().verify(password.toCharArray(), hash).verified

    private fun CustomerRow.toProfile(): CustomerProfile =
        CustomerProfile(
            id = id,
            username = username,
            name = name,
            phone = phone,
            email = email,
            avatarUrl = avatarUrl,
            signature = signature,
            isVip = isVip,
            diamondBalance = diamondBalance,
        )

    companion object {
        private const val BCRYPT_COST = 10
        private const val MIN_USERNAME_LENGTH = 3
        private const val MIN_PASSWORD_LENGTH = 6
    }
}
